src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,576 @@
+                //    They will be preserved and presented as output.
+                for mut predicate in input_knowledge.iter().cloned() {
+                    if substitute(&mut predicate, &key_exprs) {
+                        predicates.push(predicate);
+                    }
+                }
+
+                // TODO: Predicates about columns that are certain aggregates can be preserved.
+                // For example, if a column Not(IsNull) or (= 7) then Min/Max of that column
+                // will have the same property.
+                for (index, aggregate) in aggregates.iter_mut().enumerate() {
+                    let column = group_key.len() + index;
+                    if let Some(Ok(literal)) = aggregate.as_literal() {
+                        if literal == Datum::Null {
+                            predicates.push(
+                                MirScalarExpr::Column(column).call_unary(expr::UnaryFunc::IsNull),
+                            );
+                        } else {
+                            predicates.push(MirScalarExpr::Column(column).call_binary(
+                                MirScalarExpr::literal(
+                                    Ok(literal),
+                                    self_type.column_types[column].scalar_type.clone(),
+                                ),
+                                expr::BinaryFunc::Eq,
+                            ));
+                            predicates.push(
+                                MirScalarExpr::column(column)
+                                    .call_unary(UnaryFunc::IsNull)
+                                    .call_unary(UnaryFunc::Not(expr::func::Not)),
+                            );
+                        }
+                    } else {
+                        // Aggregates that are non-literals may still be non-null.
+                        if let expr::AggregateFunc::Count = aggregate.func {
+                            // Replace counts of nonnull, nondistinct expressions with `true`.
+                            if let MirScalarExpr::Column(c) = aggregate.expr {
+                                if !input_type.column_types[c].nullable && !aggregate.distinct {
+                                    aggregate.expr =
+                                        MirScalarExpr::literal_ok(Datum::True, ScalarType::Bool);
+                                }
+                            }
+                            predicates.push(
+                                MirScalarExpr::column(column)
+                                    .call_unary(UnaryFunc::IsNull)
+                                    .call_unary(UnaryFunc::Not(expr::func::Not)),
+                            );
+                        } else {
+                            // TODO: Something more sophisticated using `input_knowledge` too.
+                            if !self_type.column_types[column].nullable {
```

Comment:
This block could be moved towards the end and be common to all operators. `normalize_predicates` would get rid of any duplication introduced.